proposals/0000-modifiers.rst

```diff
@@ -0,0 +1,233 @@
+  this proposal.
+
+* Future modifiers will also seamlessly work with existing ones, where order
+  is not expected to matter (though that would be up to other proposals to
+  spell out).
+
+* Let's assume we have overloaded numbers at the type level, and then consider
+  ``%1``. Under this proposal, we would have ``1 :: a`` where ``Num a`` and
+  ``Modifier a`` must hold. If we have ``Multiplicity`` specified at the end of
+  the ambient ``default``\ing list, then ``Multiplicity`` will be the first
+  (and only) member of that list that satisfies both ``Num`` and ``Modifier``.
+  Accordingly, GHC will default ``a`` to be ``Multiplicity``, and all will be
+  well. (We may want this case to avoid activating ``-Wtype-defaults``, but
+  that's a conversation for later.)
+
+* This proposal means that ``Int %m -> Bool``, acceptable today as a
+  multiplicity-polymorphic function, would be rejected. The user would need
+  to add a kind annotation to tell us that ``m`` is a multiplicity (and not,
+  say, a matchability, which is also expected to support polymorphism). See
+  an Alternative below for a trick to mitigate this problem.
+
+* This proposal paves the way for future proposals introducing new modifiers.
+  Possible candidates:
+
+  * matchability
+  * injectivity on arrows
+  * ``oneShot``
+  * ``inline``
+  * a replacement for the ``{-# OVERLAPPING #-}`` pragmas. These pragmas
+    have, in my opinion, never really fit in: they change the semantics
+    of the declaration. Pragmas are meant to be hints or instructions
+    to the compiler, not something that changes the meaning of a program
+    and its typing rules.
+  * a mechanism for suppressing warnings over one region of a program,
+    instead of at the module level: ``%(suppress "uni-complete-patterns") (case x of ...)``.
+    This could also be done with a pragma.
+
+* Some other features that have had tortuous and torturous syntax
+  discussions may have enjoyed having the modifier option. For example,
+  this might have been used instead of ``type role`` for role annotations:
+  ``data Tagged (%Nominal t) a = Tagged a``. Or it might have been an
+  alternative for ``-XDerivingStrategies``.
+
+* Though not proposed here, we can imagine extensions allowing abstractions
+  over ``Modifiers``. This might allow being able to solve ``Modifier (a,b)``
+  when ``Modifier a`` and ``Modifier b`` holds, thus allowing something
+  like ``type ManyMatch = '(Many, Matchable); foo :: Int %ManyMatch -> Bool``.
+
+* Though not proposed here, we can imagine a large extension to this
+  mechanism allowing for *user-written* ``Modifier``\s. Perhaps a
+  ``Modifier`` type supports some function call to the GHC API that
+  transforms the meaning of bit of syntax. The possibilities are
+  tantalizing.
+  
+* These modifiers recall Java's `Annotations <https://en.wikipedia.org/wiki/Java_annotation>`_
+  mechanism, which were a direct inspiration.
```

Comment:
I think it depends on what we want the semantics of a pragma to be. I like the idea that pragmas should be out-of-band instructions to the compiler. Inlining, rewrite rules, warning enable/disable, and deprecation (among others) fit nicely here. On the other hand, annotations that change the typing or execution (beyond performance) of the program are really not out-of-band: they are an essential part of the program. These, in my opinion, should not be pragmas -- instead, these should be modifiers. The `ANN` mechanism you mention here thus seems similar to, but not the same as, the idea of modifiers.

Reply:
I think changing program semantics or not may be a useful distinction to make, but this proposal already blurs the lines by suggesting `inline` and `oneShot` as candidates for future modifiers. Neither changes the semantics of your program, only how GHC optimizes it. And it's also not super clear where `ANN` pragmas fall. They don't change the semantics on their own, but seem designed to aid plugins that may well use them to change your program's semantics!

Here's another example, I've long wanted to give LiquidHaskell refinements a more first-class feeling by teaching GHC about them to some extent. We considered using `ANN` pragmas at one point but it didn't quite work for some technical reason (I think it was surmountable, we just didn't follow through..). But would they fit better as modifiers? From GHC's perspective refinements don't change the semantics of your program (at least not at the moment), but they are an essential piece from the programmer's perspective.

Reply:
These are all really good points -- the line is blurrier than I realized. And you're right about `oneShot` and `inline`: these really should be pragmas, not modifiers. It's just that they are now magic functions, which seems worse than either.

For what it's worth, I would say that an LH annotation would be a modifier, because it changes the compile-time semantics of the program (taking a program that might have otherwise been accepted to be rejected).